File: 20-Valid-Parentheses/solution_1.py
def isValid(s):
    # build a hash "model" that contains the open and close 
    # itterate through the input and find the key value in the model and add it to the answer hash.
    # if it cannot be found see if it matches the values of any of the values in the answer hash
    # if it exists remove the pair in the answer hash
    model = {'(':')', '{': '}', '[':']'}
    answer = []
    for char in s:
        if char in model:
            answer.append(char)
        elif char in model.values():
            if not answer or model[answer.pop()] != char:
                return False
        else:
            return False
    return len(answer) == 0

File: 20-Valid-Parentheses/solution_1.js
var isValid = function (s) {
  //have an empty array/ answer stack
  //create a hashmap that have a key of opening bracket and value of closing value
  //itterate through the stack
  //set variable for each itteration of the stack
  //if the variable exists as a key in the as a map push the value into the stack
  //else if the last item in the stack is not equal to a key of the closing bracket then return false
  //in the end return a boolian for the stack length

  let stack = [];

  array = [a,b,c]
  let map = {
    '(': '[]',
    '{': '}',
    '[': ']',
  };
  for (let i = 0; i < s.length; i++) {
    let bracket = s[i];
    if (map[bracket]) {
      stack.push(map[bracket]);
    } else if (bracket !== stack.pop()) {
      return false;
    }
  }
  return !stack.length;
};
